core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java

```diff
@@ -215,14 +227,93 @@ public boolean isAvailable() {
       return file.exists();
    }
 
+   private boolean isCorruptData() throws IOException {
+      // ISPN-13128 Corrupt migration data can only be created with default marshaller
+      if (ctx.getGlobalConfiguration().serialization().marshaller() != null)
+         return false;
+
+      ByteBuffer buf = ByteBuffer.allocate(KEY_POS_LATEST);
+      long pos = filePos;
+      buf = readChannel(buf, pos, KEY_POS_LATEST);
+
+      int remaining = buf.remaining();
+      if (buf.remaining() > 0) {
+         // Remaining == KEY_POS_LATEST means the file was empty and nothing read
+         return remaining != KEY_POS_LATEST;
+      }
+      buf.flip();
+      FileEntry fe = new FileEntry(pos, buf);
+      channel.position(filePos);
+      // Explicitly calculate the correct size as incorrect value stored in fe.size
+      return fe.size != fe.actualSize();
+   }
+
+   private void migrateCorruptData() {
+      String cacheName = ctx.getCache().getName();
+      PERSISTENCE.startRecoveringCorruptPersistenceData(cacheName);
+      File newFile = new File(file.getParentFile(), cacheName + "_new.dat");
+
+      long oldFilePos = MAGIC_LATEST.length;
+      ByteBuffer buf = ByteBuffer.allocate(KEY_POS_LATEST);
+      try (FileChannel newChannel = SecurityActions.openFileChannel(newFile)) {
+         //Write Magic
+         newChannel.truncate(0);
+         newChannel.write(ByteBuffer.wrap(MAGIC_LATEST), 0);
+
+         while (true) {
+            buf = readChannel(buf, oldFilePos, KEY_POS_LATEST);
+            if (buf.remaining() > 0)
+               break;
+
+            buf.flip();
+
+            FileEntry fe = new FileEntry(oldFilePos, buf);
+            oldFilePos += KEY_POS_LATEST;
+
+            // Read old entry content and then write
+            ByRef.Long offset = new ByRef.Long(oldFilePos);
+            K key = unmarshallObject(buf, offset, fe.keyLen);
+            V value = unmarshallObject(buf, offset, fe.dataLen);
+
+            int metaLen = fe.metadataLen > 0 ? fe.metadataLen - TIMESTAMP_BYTES : 0;
+            Metadata metadata = unmarshallObject(buf, offset, metaLen);
```

Comment:
@pruivo Unmarshalling the entries actually made it more readable and easier to debug :+1: